contracts/Remittance.sol

```diff
@@ -0,0 +1,101 @@
+pragma solidity 0.5.10;
+
+import "./SafeMath.sol";
+import "./Stoppable.sol";
+
+contract Remittance is Stoppable {
+    using SafeMath for uint;
+    
+    struct Transaction {
+        uint time;
+        address emitter;
+        uint amount;
+    }
+    
+    uint constant TX_FEES = 2000;
+    uint constant CANCELLATION_DELAY = 5 days;
+    
+    mapping(bytes32 => Transaction) private transactions;
+    uint private fees;
+    
+    event LogNewTransaction(address indexed emitter, address exchange, uint amount);
+    event LogTakeFees(address indexed emitter, uint amount);
+    event LogCancelTransaction(address indexed emitter, address exchange, uint amount);
+    event LogWithdraw(address indexed emitter, uint amount);
+    
+    function hashOTP(address exchange, bytes32 password) private view returns(bytes32 hash) {
+        require(exchange != address(0) && password != 0, "Both exchange and password must be set");
+        hash = keccak256(abi.encodePacked(address(this), exchange, password));
+    }
+    
+    function newTransaction(address exchange, bytes32 password) _onlyIfRunning public payable returns(bool success) {
+        require(exchange != address(0) && password != 0 && msg.value > 0, "An error occured. Ensure that both exchange and password are set and that your transaction value is above 0");
+        
+        bytes32 hash = hashOTP(exchange, password);
```

Comment:
You know that because of how consensus works:

* all transactions are public
* all transaction data is public
* all code execution is public

What does it mean for your `password`?

Reply:
Indeed, the password can be read in cleartext. However, to be able to retrieve the transaction value, the call must be performed from the exchange wallet. What is the point in your statement?

Reply:
No `web3.eth.getTransaction(hash)`. Anyone can do that. Only access to a synchronised node is required.

Reply:
What I mean by transaction value is the possibility to withdraw the ether in the contract. What are you suggesting?

Reply:
For the exercise, Alice sends Ether for Carol to withdraw and give the $ to Bob.
We want Carol to wait for Bob and his secret to be able to withdraw.
In your setup, Carol does not need to wait because the secret is already there. Carol can withdraw without Bob's secret (not a secret).
You have to make it so that Carol has to wait for Bob for the secret.